lib/modules/dosomething/dosomething_user/dosomething_user_social_auth.inc

```diff
@@ -0,0 +1,122 @@
+<?php
+
+/**
+ * Constructs a new Drupal user for the given account details to represent a Facebook user.
+ * @param  array $account_details  All of the properties to instantiate
+ * @return user
+ */
+function dosomething_user_make_facebook_user($account_details) {
+  $edit = [
+    'mail' => $account_details['email'],
+    'name' => $account_details['email'],
+    'pass' => user_password(),
+    'status' => 1,
+    'created' => REQUEST_TIME,
+  ];
+
+  dosomething_user_set_fields($edit, $account_details);
+
+  $user = user_save('', $edit);
+  user_save($user, ['name' => $user->uid]);
+
+  return $user;
+}
+
+/**
+ * Registers a new user with Northstar (or upserts with existing) & returns a Drupal account
+ * @param  array $account_details Account details for the user to register
+ * @return user
+ */
+function dosomething_user_register_facebook_user($account_details) {
+  $user = NULL;
+
+  // Checks if there is an existing account for the given email
+  if (isset($account_details['email'])) {
+    $email = $account_details['email'];
+    $northstar_account = dosomething_northstar_get_user_by_custom_index('email', $email);
+  }
+
+  if ($northstar_account === NULL) {
+    $user = dosomething_user_make_facebook_user($account_details);
+  }
+  else {
+    $user = user_load($northstar_account->data->drupal_id);
+  }
+
+  $user->facebook_id = $account_details['facebook_id'];
+  dosomething_northstar_create_user($user);
+
+  return $user;
+}
+
+/**
+ * Verify the given access token & Facebook ID are correct with Northstar
+ * @param  string $facebook_id  Facebook ID
+ * @param  String $access_token Access token to validate
+ * @return bool
+ */
+function dosomething_user_verify_facebook_user($facebook_id, $access_token) {
+  // TODO
+  return true;
+}
+
+/**
+ * Creates a Drupal session for a given Facebook user
+ * @param $user User to create session for
+ */
+function dosomething_user_create_facebook_session($user) {
+  $form_state['uid'] = $user->uid;
+  user_login_submit([], $form_state);
+  user_login_finalize($form_state);
```

Comment:
this is for a [new user](https://github.com/DoSomething/phoenix/blob/aac08a551d112eba8b21e5ef896d5044b99ad2f5/lib/modules/dosomething/dosomething_user/dosomething_user.module#L592)